File: android/app/src/main/java/com/helix/components/TierGate.kt
/**
 * Tier Gate - Helix Android
 * Jetpack Compose component for protecting features behind subscription tiers
 * Reference: web/src/components/auth/TierGate.tsx
 */

package com.helix.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Lock
import androidx.compose.material.icons.filled.Sparkles
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.helix.core.subscription.SubscriptionService
import com.helix.core.subscription.SubscriptionTier
import com.helix.ui.theme.*

@Composable
fun TierGate(
    requiredTier: SubscriptionTier,
    subscriptionService: SubscriptionService,
    showUpgrade: Boolean = true,
    content: @Composable () -> Unit,
    upgrade: @Composable (() -> Unit)? = null
) {
    val isLoading by subscriptionService.isLoading.collectAsState()
    val hasAccess = subscriptionService.hasAccess(requiredTier)

    when {
        isLoading -> {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .background(BgPrimary),
                contentAlignment = Alignment.Center
            ) {
                CircularProgressIndicator(
                    color = HelixBlue,
                    modifier = Modifier.size(48.dp)
                )
            }
        }

        hasAccess -> {
            content()
        }

        upgrade != null -> {
            upgrade()
        }

        showUpgrade -> {
            DefaultUpgradePrompt(requiredTier)
        }
    }
}

@Composable
private fun DefaultUpgradePrompt(tier: SubscriptionTier) {
    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(BgPrimary),
        contentAlignment = Alignment.Center
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth(0.9f)
                .padding(32.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(24.dp)
        ) {
            // Lock icon
            Box(
                modifier = Modifier
                    .size(64.dp)
                    .background(BgTertiary, shape = CircleShape),
                contentAlignment = Alignment.Center
            ) {
                Icon(
                    imageVector = Icons.Default.Lock,
                    contentDescription = "Locked",
                    modifier = Modifier.size(32.dp),
                    tint = TextTertiary
                )
            }

            // Title and description
            Column(
                modifier = Modifier.fillMaxWidth(),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                Text(
                    text = "${tier.displayName} Required",
                    fontSize = 22.sp,
                    fontWeight = FontWeight.Bold,
                    color = Color.White
                )

                Text(
                    text = "This feature requires the ${tier.displayName} plan ($${tier.pricePerMonth}/month). Upgrade to unlock this and other premium features.",
                    fontSize = 16.sp,
                    color = TextSecondary,
                    textAlign = TextAlign.Center
                )
            }

            // Features list
            Card(
                modifier = Modifier.fillMaxWidth(),
                colors = CardDefaults.cardColors(
                    containerColor = BgTertiary.copy(alpha = 0.5f)
                ),
                border = BorderStroke(1.dp, Color.White.copy(alpha = 0.05f)),
                shape = RoundedCornerShape(12.dp)
            ) {
                Column(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(16.dp),
                    verticalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    // Header
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(8.dp)
                    ) {
                        Icon(
                            imageVector = Icons.Default.Sparkles,
                            contentDescription = null,
                            modifier = Modifier.size(16.dp),
                            tint = HelixAccent
                        )

                        Text(
                            text = "What you'll get:",
                            fontSize = 13.sp,
                            fontWeight = FontWeight.Medium,
                            color = TextSecondary
                        )
                    }

                    // Features
                    Column(
                        modifier = Modifier.fillMaxWidth(),
                        verticalArrangement = Arrangement.spacedBy(8.dp)
                    ) {
                        for (feature in tier.features) {
                            Row(
                                modifier = Modifier.fillMaxWidth(),
                                verticalAlignment = Alignment.CenterVertically,
                                horizontalArrangement = Arrangement.spacedBy(8.dp)
                            ) {
                                Text(
                                    text = "•",
                                    color = HelixAccent
                                )

                                Text(
                                    text = feature,
                                    fontSize = 14.sp,
                                    color = TextSecondary,
                                    modifier = Modifier.weight(1f)
                                )
                            }
                        }
                    }
                }
            }

            // Action buttons
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                Button(
                    onClick = { /* Navigate to pricing */ },
                    modifier = Modifier
                        .weight(1f)
                        .height(48.dp),
                    colors = ButtonDefaults.buttonColors(
                        containerColor = HelixBlue
                    ),
                    shape = RoundedCornerShape(8.dp)
                ) {
                    Text(
                        text = "View Plans",
                        fontSize = 16.sp,
                        fontWeight = FontWeight.SemiBold
                    )
                }

                OutlinedButton(
                    onClick = { /* Navigate to dashboard */ },
                    modifier = Modifier
                        .weight(1f)
                        .height(48.dp),
                    border = BorderStroke(1.dp, Color.White.copy(alpha = 0.1f)),
                    colors = ButtonDefaults.outlinedButtonColors(
                        containerColor = BgTertiary,
                        contentColor = TextSecondary
                    ),
                    shape = RoundedCornerShape(8.dp)
                ) {
                    Text(
                        text = "Back",
                        fontSize = 16.sp,
                        fontWeight = FontWeight.SemiBold
                    )
                }
            }
        }
    }
}

File: ios/Helix/Components/TierGate.swift
/**
 * Tier Gate - Helix iOS
 * SwiftUI view modifier for protecting features behind subscription tiers
 * Reference: web/src/components/auth/TierGate.tsx
 */

import SwiftUI

struct TierGateModifier<Upgrade: View>: ViewModifier {
    @StateObject private var subscription = SubscriptionService.shared

    let requiredTier: SubscriptionTier
    let upgrade: Upgrade?
    let showUpgrade: Bool

    func body(content: Content) -> some View {
        Group {
            if subscription.isLoading {
                ProgressView()
                    .frame(maxWidth: .infinity, maxHeight: .infinity)
                    .background(Color(hex: "0a0a0a"))
            } else if subscription.hasAccess(to: requiredTier) {
                content
            } else if let upgrade = upgrade {
                upgrade
            } else if showUpgrade {
                defaultUpgradePrompt
            } else {
                EmptyView()
            }
        }
        .task {
            await subscription.fetchSubscription()
        }
    }

    private var defaultUpgradePrompt: some View {
        VStack(spacing: 24) {
            // Icon
            ZStack {
                Circle()
                    .fill(Color(hex: "1a1a1a"))
                    .frame(width: 64, height: 64)

                Image(systemName: "lock.fill")
                    .font(.system(size: 32))
                    .foregroundColor(Color(hex: "71717a"))
            }

            VStack(spacing: 12) {
                Text("\(requiredTier.name) Required")
                    .font(.system(size: 22, weight: .bold))
                    .foregroundColor(.white)

                Text("This feature requires the \(requiredTier.name) plan ($\(requiredTier.pricePerMonth)/month). Upgrade to unlock this and other premium features.")
                    .font(.system(size: 16))
                    .foregroundColor(Color(hex: "a1a1aa"))
                    .multilineTextAlignment(.center)
            }

            // Features list
            VStack(spacing: 8) {
                HStack(spacing: 8) {
                    Image(systemName: "sparkles")
                        .font(.system(size: 12))
                        .foregroundColor(Color(hex: "33a7e7"))

                    Text("What you'll get:")
                        .font(.system(size: 13, weight: .medium))
                        .foregroundColor(Color(hex: "d4d4d8"))
                }

                VStack(spacing: 8) {
                    ForEach(requiredTier.features, id: \.self) { feature in
                        HStack(spacing: 8) {
                            Text("•")
                                .foregroundColor(Color(hex: "33a7e7"))

                            Text(feature)
                                .font(.system(size: 14))
                                .foregroundColor(Color(hex: "a1a1aa"))

                            Spacer()
                        }
                    }
                }
            }
            .padding(16)
            .background(Color(hex: "1a1a1a").opacity(0.5))
            .cornerRadius(12)
            .overlay(
                RoundedRectangle(cornerRadius: 12)
                    .stroke(Color(hex: "27272a"), lineWidth: 1)
            )

            HStack(spacing: 12) {
                NavigationLink(destination: Text("Pricing Page")) {
                    Text("View Plans")
                        .font(.system(size: 16, weight: .semibold))
                        .foregroundColor(.white)
                        .frame(maxWidth: .infinity)
                        .padding(12)
                        .background(Color(hex: "0686D4"))
                        .cornerRadius(8)
                }

                NavigationLink(destination: Text("Dashboard")) {
                    Text("Back to Dashboard")
                        .font(.system(size: 16, weight: .semibold))
                        .foregroundColor(Color(hex: "a1a1aa"))
                        .frame(maxWidth: .infinity)
                        .padding(12)
                        .background(Color(hex: "1a1a1a"))
                        .cornerRadius(8)
                }
            }

            Spacer()
        }
        .padding(32)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(Color(hex: "0a0a0a"))
    }
}

extension View {
    /// Gate a view behind a subscription tier
    func tierGate(
        requiredTier: SubscriptionTier,
        showUpgrade: Bool = true,
        upgrade: AnyView? = nil
    ) -> some View {
        self.modifier(TierGateModifier(requiredTier: requiredTier, upgrade: upgrade, showUpgrade: showUpgrade))
    }

    /// Gate a view with custom upgrade view
    func tierGate<Upgrade: View>(
        requiredTier: SubscriptionTier,
        @ViewBuilder upgrade: () -> Upgrade
    ) -> some View {
        self.modifier(TierGateModifier(requiredTier: requiredTier, upgrade: upgrade(), showUpgrade: false))
    }
}

// MARK: - Preview

#if DEBUG
struct TierGate_Previews: PreviewProvider {
    static var previews: some View {
        VStack {
            Text("Architect Feature")
                .font(.title)
        }
        .tierGate(requiredTier: .architect)
    }
}
#endif
